### PR TITLE
Support Latex in Visdom

### PR DESCRIPTION
**Pull request by ubaidsworld**
Visdom will support the latex.
## Description

The MathJax Library has some issues while working locally. Whereas using the cdnjs cloud, it was working well.
Hence, introduced the MathJax in the static js folder which will directly get installed while installing Visdom.
## Motivation and Context

Had seen many queries and found how much it a need for the people using Visdom for their project.
Hence, thought of helping everyone in need of this feature.
The following issues will get resolved: 
#723 

## How Has This Been Tested?

Yes, I have tested it and have updated the demo.py.
Here is an example and the screenshot of the output is added accordingly.
```
import visdom
import numpy as np
viz = visdom.Visdom()

viz.line(
	X=[1, 2, 3, 4],
	Y=[1, 4, 9, 16],
	win="latex_support",
	name=r'$\alpha_{1c} = 352 \pm 11 \text{ km s}^{-1}$',
	update='append',
	opts=dict(
	showlegend=True,
	title="Demo Latex in Visdom",
	xlabel=r'$\sqrt{(n_\text{c}(t|{T_\text{early}}))}$',
	ylabel=r'$d, r \text{ (solar radius)}$',
	)
)
viz.line(
	X=[1, 2, 3, 4],
	Y=[0.5, 2, 4.5, 8],
	win="latex_support",
	name=r'$\beta_{1c} = 25 \pm 11 \text{ km s}^{-1}$',
	update='append',
)
```
## Screenshots (if appropriate):
Latex Support

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] For JavaScript changes, I have re-generated the minified JavaScript code.

### CODE REVIEW

**Comment by ubaidsworld** [comment]
Hello @JackUrb, now the issue is resolve and you can verify it. I hope this time you won't find issues.

**Comment by ubaidsworld** [comment]
Hello, please let me know if it is worth merging. It will definitely solve the latex issue.

File: MANIFEST.in [hunk 1]
```diff
@@ -2,6 +2,6 @@ include README.md
 include py/visdom/VERSION
 include py/visdom/py.typed
 include py/visdom/*.pyi
-recursive-include py/visdom/static/*
+recursive-include py/visdom/static *
```
Comment (JackUrb) [inline MANIFEST.in]: What is the purpose of this change?

File: README.md [hunk 2]
```diff
@@ -714,6 +714,9 @@ Arguments:
 ## License
 visdom is Creative Commons Attribution-NonCommercial 4.0 International Public licensed, as found in the LICENSE file.
 
+## Note on Latex Support
+Now visdom support Latex. Check the examples/demo.py.
+
```
Comment (JackUrb) [inline README.md]: Generally we use the demo to showcase possible functionality in visdom, so your addition to the demo is great and this note isn't necessary.

File: example/demo.py [hunk 3]
```diff
@@ -485,6 +485,28 @@ def img_click_callback(event):
     Y2 = np.random.randint(0, 20, 20)
     viz.dual_axis_lines(X, Y1, Y2)
 
+    # Example for Latex Support
+    viz.line(
```
Comment (JackUrb) [inline example/demo.py:489]: This is great!

File: example/demo.py [hunk 4]
```diff
@@ -485,6 +485,28 @@ def img_click_callback(event):
...
+        Y=[1, 4, 9, 16],
+        win="latex_support",
+        name=r'$\alpha_{1c} = 352 \pm 11 \text{ km s}^{-1}$',
+        update='append',
+        opts=dict(
+        showlegend=True,
+        title="Demo Latex in Visdom",
+        xlabel=r'$\sqrt{(n_\text{c}(t|{T_\text{early}}))}$',
+        ylabel=r'$d, r \text{ (solar radius)}$',
+        )
```
Comment (JackUrb) [inline example/demo.py]: Use `{'key': value}` formatting for dicts.

File: py/visdom/static/index.html [hunk 5]
```diff
@@ -30,7 +30,10 @@
     <script src={{ static_url("js/layout_bin_packer.js") }}></script>
 
     <!-- Mathjax -->
-    <script type="text/javascript" async src={{ static_url("js/mathjax-MathJax.js") }}></script>
+   	<script type='text/javascript' src={{ static_url('js/MathJax/MathJax.js') }}></script>
+    <script type='text/x-mathjax-config'>MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']], processEscapes: true}});</script>
+    <script type='text/javascript' src={{ static_url('js/MathJax/config/TeX-AMS-MML_HTMLorMML.js') }}></script>
```
Comment (JackUrb) [inline py/visdom/static/index.html]: Nice

File: py/visdom/server.py [hunk 6]
```diff
@@ -1653,8 +1653,6 @@ def download_scripts(proxies=None, install_dir=None):
             'react-dom.min.js',
         '%sreact-modal@3.1.10/dist/react-modal.min.js' % b:
             'react-modal.min.js',
-        'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_SVG':  # noqa
-            'mathjax-MathJax.js',
```
Comment (JackUrb) [inline py/visdom/server.py:1657]: Rather than removing this download entirely and committing the MathJax files to the `Visdom` repo, you should add extra download paths here to create and fill out the `static/js/MathJax` folder with the files you've found are required to get everything to work.

**Comment by ubaidsworld** [comment]
> 
> 
> Nice work here @ubaidsworld, you've found the full tree of required files to get MathJax working and also added a great demo.
> 
> The one constraint we have though is that we're not allowed to commit the external MathJax source to our repository, so you'll instead need to update the download script to save the files you've found necessary to the paths you've noted here. This way on running `server.py`'s `download_scripts()` function, the `static/js/MathJax` directory will have all of the files that you've currently committed here.

Actually, I thought it would be better that I download the file and keep it in the necessary directory (static/js). So that way we can let people install everything when the Visdom installs. But if that is a protocol then I will try make it in the `server.py`'s `download_scripts()`.
It will take a little while as I am on something and might be doing it till this weekend. 
Even was trying to add the replace option in the `dual_axis_line()` feature, but have other priorities though.
Well, I will make it if it is not doable this way.

File: MANIFEST.in (hunk 1 above)
Comment (ubaidsworld) [reply MANIFEST.in]: Actually, when I tried using the former syntax and run the command `python setup.py install` to test, MathJax folder was missing in the js. Hence, researched a bit and found something what the latter statements, and _hallelujah_, everything worked. But as you have mentioned that you will not allow just commit, then this change can be revoked.
Still, this change won't harm the sequence for sure.
I have tested it.

File: README.md (hunk 2 above)
Comment (ubaidsworld) [reply README.md]: Ok I will remove the README.md changed from the Pull Request.

File: example/demo.py (hunk 3 above)
Comment (ubaidsworld) [reply example/demo.py:489]: Thank you for appreciation... :)

File: example/demo.py (hunk 4 above)
Comment (ubaidsworld) [reply example/demo.py]: oops.. I missed it again, I will make it proper.

File: py/visdom/static/index.html (hunk 5 above)
Comment (ubaidsworld) [reply py/visdom/static/index.html]: Thank you once again..... :)

File: py/visdom/server.py (hunk 6 above)
Comment (ubaidsworld) [reply py/visdom/server.py:1657]: Ok I will try it out. But may be I need to write another code for that. I will try using the available one. But if not then I ill write my own code there and it will not break other cod, for sure.

**Comment by ubaidsworld** [comment]
Thank you for your review, @JackUrb .

**Comment by ubaidsworld** [comment]
Hi @JackUrb ,
Check the new updates. I hope this will help and you will merge this pull request.

File: py/visdom/server.py (hunk 6 above)
Comment (ubaidsworld) [reply py/visdom/server.py:1657]: I have added the code

File: example/demo.py (hunk 4 above)
Comment (ubaidsworld) [reply example/demo.py]: Changes Done...:)

**Comment by ubaidsworld** [comment]
Hi @JackUrb ,
Check the new updates. I hope this will help and you will merge this pull request.

File: py/visdom/server.py [hunk 7]
```diff
@@ -1743,6 +1742,32 @@ def download_scripts(proxies=None, install_dir=None):
                 logging.error('Error {} while downloading {}'.format(
                     exc.reason, key))
 
+    # Download MathJax Js Files
+    import requests
+    cdnjs_url = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/' 
+    mathjax_dir='\\'.join(cdnjs_url.split('/')[-3:])
```
Comment (JackUrb) [inline py/visdom/server.py]: I'm not sure this is the most effective syntax for what you're trying to do here - to get the os specific path, you can use `os.path.join(*cdnjs_url.split('/')[-3:])`. In general you should avoid assuming details about the os and pathing (which is what I _think_ you're doing with `'\\'`).

Comment (ubaidsworld) [reply py/visdom/server.py]: I did not understood this comment.
The line: `mathjax_dir='\\'.join(cdnjs_url.split('/')[-3:])` will return the value as: mathjax\\2.7.5. 
But if that is not good syntax then I will replace it with `os.path.join()`

Comment (ubaidsworld) [reply py/visdom/server.py]: Changes Done!.....:)

Comment (JackUrb) [reply py/visdom/server.py]: `mathjax\2.7.5` is the directory structure only for windows machines. Unix machines use `/` as the directory separator, so we use `os.path.join` rather than directly joining with a character, as the `os.path` library determines whether to use `\` or `/` based on the platform running the script.

File: py/visdom/server.py [hunk 8]
```diff
@@ -1743,6 +1742,32 @@ def download_scripts(proxies=None, install_dir=None):
...
+        'jax/output/SVG/jax.js?V=2.7.5',
+        'jax/output/SVG/fonts/TeX/Size1/Regular/Main.js?V=2.7.5',
+        'jax/output/SVG/config.js?V=2.7.5',
+        'MathJax.js?config=TeX-AMS-MML_HTMLorMML%2CSafe.js&#038;ver=4.1',
+    ]
+    mathjax_dir_path = '%s/static/%s/%s' % (install_dir, 'js',mathjax_dir)    
+    
+    for path in mathjax_path:
+        filename=path.split("/")[-1].split("?")[0]
+        extracted_directory=os.path.join(mathjax_dir_path,"\\".join(path.split('/')[:-1]))
```
Comment (JackUrb) [inline py/visdom/server.py]: As a note for good OSS contributing practice, I won't always point out every location that has an issue if I've noted it elsewhere in the file. `"\\".join(<list>)` should be `os.path.join(*<list>)` as above.

Comment (ubaidsworld) [reply py/visdom/server.py]: oops.. I am really sorry. I neglected it. Thank you. I will check it out.

Comment (ubaidsworld) [reply py/visdom/server.py]: There are no `'\\'.join(<list>)` now.

**Comment by ubaidsworld** [comment]
> 
> 
> Thanks for all the work on making sure this is good to go! This should be ready to pull in and close one of our longer-standing bugs 👍

All thanks to you @JackUrb. I have learned many things while working with visdom.

**Comment by ubaidsworld** [comment]
@JackUrb Why the request is not merged even after being approved?

**Comment by JackUrb** [comment]
As in last time, it has to undergo internal approval, where additional changes may be made

**Comment by ubaidsworld** [comment]
Ah.. okay. I got it. Thank you for the clarification. :)
